File: ios/Classes/NeiryKitPlugin.swift
import Flutter
import UIKit

public class NeiryKitPlugin: NSObject, FlutterPlugin {
    private static var instance: NeiryKitPlugin?

    private let registrar: FlutterPluginRegistrar
    private var methodChannels: [String: FlutterMethodChannel] = [:]
    private var eventChannels: [String: FlutterEventChannel] = [:]
    private var deviceLocatorBridge: DeviceLocatorBridge?
    private var deviceBridge: DeviceBridge?
    private var nfbBridge: NfbBridge?
    private var nfbCalibratorBridge: NfbCalibratorBridge?
    private var emotionsBridge: EmotionsBridge?
    private var physioBridge: PhysioBridge?
    private var cardioBridge: CardioBridge?
    private var productivityBridge: ProductivityBridge?

    private init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = NeiryKitPlugin(registrar: registrar)
        self.instance = instance
        instance.registerMethodChannels()
        instance.deviceLocatorBridge = DeviceLocatorBridge()
        instance.deviceBridge = DeviceBridge()
        instance.nfbBridge = NfbBridge()
        instance.nfbCalibratorBridge = NfbCalibratorBridge()
        instance.emotionsBridge = EmotionsBridge()
        instance.physioBridge = PhysioBridge()
        instance.cardioBridge = CardioBridge()
        instance.productivityBridge = ProductivityBridge()
        instance.registerEventChannels()
    }

    // MARK: - MethodChannel registration

    private func registerMethodChannels() {
        let messenger = registrar.messenger()
        let ids: [String] = [
            "neiry_kit/device_locator",
            "neiry_kit/device",
            "neiry_kit/nfb",
            "neiry_kit/physiological",
            "neiry_kit/emotions",
            "neiry_kit/productivity",
            "neiry_kit/cardio",
            "neiry_kit/nfb_calibrator",
        ]
        for id in ids {
            let channel = FlutterMethodChannel(name: id, binaryMessenger: messenger)
            methodChannels[id] = channel
            channel.setMethodCallHandler { [weak self] call, result in
                self?.handleMethodCall(call, result: result, channelId: id)
            }
        }
    }

    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult, channelId: String) {
        if channelId == "neiry_kit/device_locator" {
            handleDeviceLocatorCall(call, result: result)
        } else if channelId == "neiry_kit/device" {
            handleDeviceCall(call, result: result)
        } else if channelId == "neiry_kit/nfb" {
            handleNfbCall(call, result: result)
        } else if channelId == "neiry_kit/nfb_calibrator" {
            handleNfbCalibratorCall(call, result: result)
        } else if channelId == "neiry_kit/emotions" {
            handleEmotionsCall(call, result: result)
        } else if channelId == "neiry_kit/physiological" {
            handlePhysiologicalCall(call, result: result)
        } else if channelId == "neiry_kit/cardio" {
            handleCardioCall(call, result: result)
        } else if channelId == "neiry_kit/productivity" {
            handleProductivityCall(call, result: result)
        } else {
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - device_locator dispatch

    private func handleDeviceLocatorCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let bridge = deviceLocatorBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "DeviceLocatorBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "getVersionString":
            let version = String(cString: clCCapsule_GetVersionString())
            result(version)
        case "create":
            let logDirectory = (call.arguments as? [String: Any])?["logDirectory"] as? String
            do {
                try bridge.create(logDirectory: logDirectory)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "createDevice":
            guard let args = call.arguments as? [String: Any],
                  let serial = args["serial"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'serial'", details: nil))
                return
            }
            do {
                let ret = try bridge.createDevice(serial: serial)
                if let handle = DeviceLocatorBridge.devices[serial] {
                    deviceBridge?.setDevice(serial: serial, handle: handle)
                    DeviceLocatorBridge.devices.removeValue(forKey: serial)
                }
                result(ret)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "setSingleThreaded":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'enabled'", details: nil))
                return
            }
            bridge.setSingleThreaded(enabled: enabled)
            result(nil)
        case "update":
            do {
                try bridge.update()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "setLogLevel":
            guard let args = call.arguments as? [String: Any],
                  let level = args["level"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'level'", details: nil))
                return
            }
            bridge.setLogLevel(level: level)
            result(nil)
        case "dispose":
            nfbCalibratorBridge?.stopCalibration()
            cardioBridge?.dispose()
            productivityBridge?.dispose()
            emotionsBridge?.dispose()
            physioBridge?.dispose()
            nfbBridge?.dispose()
            bridge.dispose()
            deviceBridge?.release()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - device dispatch

    private func handleDeviceCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let bridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "DeviceBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "connect":
            do {
                try bridge.connect(call)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "disconnect":
            do {
                try bridge.disconnect()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "start":
            do {
                try bridge.start()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "stop":
            do {
                let ret = try bridge.stop()
                result(ret)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getInfo":
            do {
                let info = try bridge.getInfo()
                result(info)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getBatteryCharge":
            do {
                let charge = try bridge.getBatteryCharge()
                result(charge)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getMode":
            result(bridge.getMode())
        case "getEEGSampleRate":
            do {
                let rate = try bridge.getEEGSampleRate()
                result(rate)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getPPGSampleRate":
            do {
                let rate = try bridge.getPPGSampleRate()
                result(rate)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getMEMSSampleRate":
            do {
                let rate = try bridge.getMEMSSampleRate()
                result(rate)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getPPGIrAmplitude":
            do {
                let amplitude = try bridge.getPPGIrAmplitude()
                result(amplitude)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getPPGRedAmplitude":
            do {
                let amplitude = try bridge.getPPGRedAmplitude()
                result(amplitude)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getChannelNames":
            do {
                let names = try bridge.getChannelNames()
                result(names)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getChannelsCount":
            do {
                let count = try bridge.getChannelsCount()
                result(count)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getChannelIndexByName":
            do {
                let index = try bridge.getChannelIndexByName(call)
                result(index)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getChannelNameByIndex":
            do {
                let name = try bridge.getChannelNameByIndex(call)
                result(name)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getFirmwareVersion":
            do {
                let version = try bridge.getFirmwareVersion()
                result(version)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - nfb dispatch

    private func handleNfbCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let nfbBridge = nfbBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "NfbBridge or DeviceBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "create":
            do {
                let dev = try deviceBridge.requireDevice()
                try nfbBridge.create(device: dev)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "createCalibrated":
            let args = call.arguments as? [String: Any]
            let calibrationData = args?["calibrationData"] as? [String: Any]
            do {
                let dev = try deviceBridge.requireDevice()
                try nfbBridge.createCalibrated(device: dev, calibrationData: calibrationData)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "dispose":
            nfbBridge.dispose()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - nfb_calibrator dispatch

    private func handleNfbCalibratorCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let nfbCalibratorBridge = nfbCalibratorBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED",
                                message: "NfbCalibratorBridge or DeviceBridge not initialized",
                                details: nil))
            return
        }
        switch call.method {
        case "startCalibration":
            let args = call.arguments as? [String: Any]
            let quick = (args?["calibratorData"] as? String) == "quick"
            do {
                let dev = try deviceBridge.requireDevice()
                try nfbCalibratorBridge.startCalibration(device: dev, quick: quick, result: result)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "stopCalibration":
            nfbCalibratorBridge.stopCalibration()
            result(nil)
        case "importCalibration":
            guard let args = call.arguments as? [String: Any],
                  let calibratorData = args["calibratorData"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'calibratorData'", details: nil))
                return
            }
            do {
                let dev = try deviceBridge.requireDevice()
                try nfbCalibratorBridge.importCalibration(device: dev, calibratorData: calibratorData)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "getCalibration":
            do {
                let dev = try deviceBridge.requireDevice()
                let map = try nfbCalibratorBridge.getCalibration(device: dev)
                result(map)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "isCalibrated":
            do {
                let dev = try deviceBridge.requireDevice()
                result(nfbCalibratorBridge.isCalibrated(device: dev))
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - emotions dispatch

    private func handleEmotionsCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let emotionsBridge = emotionsBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "EmotionsBridge or DeviceBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "create":
            do {
                let dev = try deviceBridge.requireDevice()
                try emotionsBridge.create(device: dev)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "dispose":
            emotionsBridge.dispose()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - physiological dispatch

    private func handlePhysiologicalCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let physioBridge = physioBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "PhysioBridge or DeviceBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "create":
            do {
                let dev = try deviceBridge.requireDevice()
                try physioBridge.create(device: dev)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "startBaselineCalibration":
            do {
                try physioBridge.startBaselineCalibration()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "importBaselines":
            let args = call.arguments as? [String: Any]
            guard let baselines = args?["baselines"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'baselines'", details: nil))
                return
            }
            do {
                try physioBridge.importBaselines(map: baselines)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "dispose":
            physioBridge.dispose()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - cardio dispatch

    private func handleCardioCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let cardioBridge = cardioBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "CardioBridge or DeviceBridge not initialized", details: nil))
            return
        }
        switch call.method {
        case "create":
            do {
                let dev = try deviceBridge.requireDevice()
                try cardioBridge.create(device: dev)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "createCalibrated":
            let args = call.arguments as? [String: Any]
            let calibrationData = args?["calibrationData"] as? [String: Any]
            do {
                let dev = try deviceBridge.requireDevice()
                try cardioBridge.createCalibrated(device: dev, calibrationData: calibrationData)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "dispose":
            cardioBridge.dispose()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - productivity dispatch

    private func handleProductivityCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let productivityBridge = productivityBridge, let deviceBridge = deviceBridge else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "ProductivityBridge or DeviceBridge not initialized", details: nil))
            return
        }
        let args = call.arguments as? [String: Any]
        switch call.method {
        case "create":
            do {
                let dev = try deviceBridge.requireDevice()
                try productivityBridge.create(device: dev)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "createCalibrated":
            guard let calibrationData = args?["calibrationData"] as? [String: Any] else {
                result(FlutterError(code: "INVALID_ARGS", message: "calibrationData is required", details: nil))
                return
            }
            do {
                let dev = try deviceBridge.requireDevice()
                try productivityBridge.createCalibrated(device: dev, calibrationData: calibrationData)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "startBaselineCalibration":
            do {
                try productivityBridge.startBaselineCalibration()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "importBaselines":
            guard let baselines = args?["baselines"] as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing 'baselines'", details: nil))
                return
            }
            do {
                try productivityBridge.importBaselines(data: baselines)
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "resetAccumulatedFatigue":
            do {
                try productivityBridge.resetAccumulatedFatigue()
                result(nil)
            } catch let e as FlutterError {
                result(e)
            } catch {
                result(FlutterError(code: "UNKNOWN", message: "\(error)", details: nil))
            }
        case "dispose":
            productivityBridge.dispose()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - EventChannel registration

    private func registerEventChannels() {
        let messenger = registrar.messenger()

        // Build lookup for the 8 live device stream handlers
        var deviceHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = deviceBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                deviceHandlers[id] = handler
            }
        }

        // Build lookup for the 2 NFB stream handlers
        var nfbHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = nfbBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                nfbHandlers[id] = handler
            }
        }

        // Build lookup for the NFB calibration stream handler
        var nfbCalibratorHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = nfbCalibratorBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                nfbCalibratorHandlers[id] = handler
            }
        }

        // Build lookup for the 2 Emotions stream handlers
        var emotionsHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = emotionsBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                emotionsHandlers[id] = handler
            }
        }

        // Build lookup for the 4 Physiological stream handlers
        var physioHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = physioBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                physioHandlers[id] = handler
            }
        }

        // Build lookup for the 3 Cardio stream handlers
        var cardioHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = cardioBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                cardioHandlers[id] = handler
            }
        }

        // Build lookup for the 6 Productivity stream handlers
        var productivityHandlers: [String: FlutterStreamHandler] = [:]
        if let bridge = productivityBridge {
            for (id, handler) in bridge.allStreamHandlers() {
                productivityHandlers[id] = handler
            }
        }

        let ids: [String] = [
            "neiry_kit/events/deviceList",
            "neiry_kit/events/eeg",
            "neiry_kit/events/psd",
            "neiry_kit/events/eegArtifacts",
            "neiry_kit/events/resistance",
            "neiry_kit/events/battery",
            "neiry_kit/events/connectionStatus",
            "neiry_kit/events/modeSwitched",
            "neiry_kit/events/nfbState",
            "neiry_kit/events/physiologicalState",
            "neiry_kit/events/emotionsState",
            "neiry_kit/events/productivityMetrics",
            "neiry_kit/events/productivityIndexes",
            "neiry_kit/events/cardioData",
            "neiry_kit/events/ppgData",
            "neiry_kit/events/memsData",
            "neiry_kit/events/nfbCalibration",
            "neiry_kit/events/physiologicalCalibrationProgress",
            "neiry_kit/events/physiologicalCalibrated",
            "neiry_kit/events/physiologicalIndividualNfb",
            "neiry_kit/events/productivityCalibrationProgress",
            "neiry_kit/events/productivityCalibrated",
            "neiry_kit/events/productivityBaselines",
            "neiry_kit/events/productivityIndividualNfb",
            "neiry_kit/events/cardioCalibratedEvent",
            "neiry_kit/events/error",
            "neiry_kit/events/nfbError",
            "neiry_kit/events/emotionsError",
            "neiry_kit/events/productivityError",
        ]
        for id in ids {
            let channel = FlutterEventChannel(name: id, binaryMessenger: messenger)
            if id == "neiry_kit/events/deviceList" {
                channel.setStreamHandler(deviceLocatorBridge)
            } else if let handler = deviceHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = nfbHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = nfbCalibratorHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = emotionsHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = physioHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = cardioHandlers[id] {
                channel.setStreamHandler(handler)
            } else if let handler = productivityHandlers[id] {
                channel.setStreamHandler(handler)
            } else {
                channel.setStreamHandler(StubStreamHandler())
            }
            eventChannels[id] = channel
        }
    }
}

// MARK: - StubStreamHandler

private class StubStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

File: android/src/main/kotlin/com/neiry/neiry_kit/NeiryKitPlugin.kt
package com.neiry.neiry_kit

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** NeiryKitPlugin */
class NeiryKitPlugin : FlutterPlugin, MethodCallHandler {

    private var methodChannels: MutableMap<String, MethodChannel> = mutableMapOf()
    private var eventChannels: MutableMap<String, EventChannel> = mutableMapOf()

    private var nativeBridge: NativeBridge? = null
    private var deviceLocatorBridge: DeviceLocatorBridge? = null
    private var deviceBridge: DeviceBridge? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        nativeBridge = NativeBridge()  // triggers System.loadLibrary via companion init
        deviceLocatorBridge = DeviceLocatorBridge(nativeBridge!!, Handler(Looper.getMainLooper()))
        deviceBridge = DeviceBridge(nativeBridge!!)

        val messenger = flutterPluginBinding.binaryMessenger

        // Register all MethodChannels
        val methodChannelIds = listOf(
            "neiry_kit/device_locator",
            "neiry_kit/device",
            "neiry_kit/nfb",
            "neiry_kit/physiological",
            "neiry_kit/emotions",
            "neiry_kit/productivity",
            "neiry_kit/cardio",
            "neiry_kit/nfb_calibrator",
        )
        for (id in methodChannelIds) {
            val channel = MethodChannel(messenger, id)
            channel.setMethodCallHandler { call, result ->
                handleMethodCall(call, result, id)
            }
            methodChannels[id] = channel
        }

        // Register all EventChannels — deviceList uses the real bridge, others use stub
        val eventChannelIds = listOf(
            "neiry_kit/events/deviceList",
            "neiry_kit/events/eeg",
            "neiry_kit/events/psd",
            "neiry_kit/events/eegArtifacts",
            "neiry_kit/events/resistance",
            "neiry_kit/events/battery",
            "neiry_kit/events/connectionStatus",
            "neiry_kit/events/modeSwitched",
            "neiry_kit/events/nfbState",
            "neiry_kit/events/physiologicalState",
            "neiry_kit/events/emotionsState",
            "neiry_kit/events/productivityMetrics",
            "neiry_kit/events/productivityIndexes",
            "neiry_kit/events/cardioData",
            "neiry_kit/events/ppgData",
            "neiry_kit/events/memsData",
            "neiry_kit/events/nfbCalibration",
            "neiry_kit/events/physiologicalCalibrationProgress",
            "neiry_kit/events/physiologicalCalibrated",
            "neiry_kit/events/physiologicalIndividualNfb",
            "neiry_kit/events/productivityCalibrationProgress",
            "neiry_kit/events/productivityCalibrated",
            "neiry_kit/events/productivityBaselines",
            "neiry_kit/events/productivityIndividualNfb",
            "neiry_kit/events/cardioCalibratedEvent",
            "neiry_kit/events/error",
            "neiry_kit/events/nfbError",
            "neiry_kit/events/emotionsError",
            "neiry_kit/events/productivityError",
        )
        for (id in eventChannelIds) {
            val channel = EventChannel(messenger, id)
            val handler: EventChannel.StreamHandler = when (id) {
                "neiry_kit/events/deviceList" -> deviceLocatorBridge!!
                else -> StubStreamHandler()
            }
            channel.setStreamHandler(handler)
            eventChannels[id] = channel
        }
    }

    private fun handleMethodCall(call: MethodCall, result: Result, channelId: String) {
        when (channelId) {
            "neiry_kit/device_locator" -> handleDeviceLocatorCall(call, result)
            "neiry_kit/device"         -> handleDeviceCall(call, result)
            "neiry_kit/nfb"            -> handleNfbCall(call, result)
            "neiry_kit/nfb_calibrator" -> handleNfbCalibratorCall(call, result)
            "neiry_kit/emotions"       -> handleEmotionsCall(call, result)
            "neiry_kit/physiological"  -> handlePhysiologicalCall(call, result)
            "neiry_kit/cardio"         -> handleCardioCall(call, result)
            "neiry_kit/productivity"   -> handleProductivityCall(call, result)
            else                       -> result.notImplemented()
        }
    }

    private fun handleDeviceLocatorCall(call: MethodCall, result: Result) {
        val bridge = deviceLocatorBridge
            ?: return result.error("NOT_INITIALIZED", "DeviceLocatorBridge not initialized", null)
        try {
            when (call.method) {
                "getVersionString" -> result.success(nativeBridge!!.nativeGetVersion())

                "create" -> {
                    val logDirectory = (call.arguments as? Map<*, *>)?.get("logDirectory") as? String
                    bridge.create(logDirectory)
                    result.success(null)
                }

                "createDevice" -> {
                    val serial = (call.arguments as? Map<*, *>)?.get("serial") as? String
                        ?: return result.error("INVALID_ARGS", "Missing 'serial'", null)
                    bridge.createDevice(serial)
                    val deviceHandle = DeviceLocatorBridge.devices[serial]
                    if (deviceHandle != null) {
                        deviceBridge?.setDevice(serial, deviceHandle)
                        DeviceLocatorBridge.devices.remove(serial)
                    }
                    result.success(null)
                }

                "setSingleThreaded" -> {
                    val enabled = (call.arguments as? Map<*, *>)?.get("enabled") as? Boolean
                        ?: return result.error("INVALID_ARGS", "Missing 'enabled'", null)
                    bridge.setSingleThreaded(enabled)
                    result.success(null)
                }

                "update" -> {
                    bridge.update()
                    result.success(null)
                }

                "setLogLevel" -> {
                    val level = (call.arguments as? Map<*, *>)?.get("level") as? Int
                        ?: return result.error("INVALID_ARGS", "Missing 'level'", null)
                    bridge.setLogLevel(level)
                    result.success(null)
                }

                "dispose" -> {
                    bridge.dispose()
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        } catch (e: FlutterError) {
            result.error(e.code, e.message, e.details)
        } catch (e: Exception) {
            result.error("UNKNOWN", e.message, null)
        }
    }

    private fun handleDeviceCall(call: MethodCall, result: Result) {
        val bridge = deviceBridge
            ?: return result.error("NOT_INITIALIZED", "DeviceBridge not initialized", null)
        try {
            when (call.method) {
                "connect" -> {
                    val bipolarChannels = (call.arguments as? Map<*, *>)?.get("bipolarChannels") as? Boolean ?: false
                    bridge.connect(bipolarChannels)
                    result.success(null)
                }
                "disconnect" -> {
                    bridge.disconnect()
                    result.success(null)
                }
                "start" -> {
                    bridge.start()
                    result.success(null)
                }
                "stop" -> result.success(bridge.stop())
                "getInfo" -> result.notImplemented()
                "getBatteryCharge" -> result.success(bridge.getBatteryCharge())
                "getMode" -> result.success(bridge.getMode())
                "getEEGSampleRate" -> result.success(bridge.getEEGSampleRate())
                "getPPGSampleRate" -> result.success(bridge.getPPGSampleRate())
                "getMEMSSampleRate" -> result.success(bridge.getMEMSSampleRate())
                "getPPGIrAmplitude" -> result.success(bridge.getPPGIrAmplitude())
                "getPPGRedAmplitude" -> result.success(bridge.getPPGRedAmplitude())
                "getChannelNames" -> result.success(bridge.getChannelNames())
                "getChannelsCount" -> result.success(bridge.getChannelsCount())
                "getChannelIndexByName" -> {
                    val channelName = (call.arguments as? Map<*, *>)?.get("channelName") as? String
                        ?: return result.error("INVALID_ARGS", "Missing 'channelName'", null)
                    result.success(bridge.getChannelIndexByName(channelName))
                }
                "getChannelNameByIndex" -> {
                    val index = (call.arguments as? Map<*, *>)?.get("index") as? Int
                        ?: return result.error("INVALID_ARGS", "Missing 'index'", null)
                    result.success(bridge.getChannelNameByIndex(index))
                }
                else -> result.notImplemented()
            }
        } catch (e: FlutterError) {
            result.error(e.code, e.message, e.details)
        } catch (e: Exception) {
            result.error("UNKNOWN", e.message, null)
        }
    }

    private fun handleNfbCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    private fun handleNfbCalibratorCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    private fun handleEmotionsCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    private fun handlePhysiologicalCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    private fun handleCardioCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    private fun handleProductivityCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        result.notImplemented()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        deviceLocatorBridge?.dispose()
        deviceBridge?.release()
        deviceLocatorBridge = null
        deviceBridge = null
        nativeBridge = null

        for ((_, channel) in methodChannels) {
            channel.setMethodCallHandler(null)
        }
        methodChannels.clear()

        for ((_, channel) in eventChannels) {
            channel.setStreamHandler(null)
        }
        eventChannels.clear()
    }

    // Stub stream handler — no-op onListen/onCancel, mirrors iOS StubStreamHandler
    private inner class StubStreamHandler : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {}
        override fun onCancel(arguments: Any?) {}
    }
}
